### ISSUE TITLE
Idea: Cyberware

### ISSUE THREAD

**Issue by AmkG** [jwvhewitt/gearhead-caramel#78]
Cyberware was one of my favorite mechanics in GearHead 1, so I'm wondering if it is planned for Caramel as well, or would it be out-of-scope because of the "coffeebreaklike" nature?

In keeping with removing the weird randomness of Mecha Engineering and opening up mecha design to all characters without requiring a skill, we could open up Cyberware to all characters without requiring a separate skill.

Each piece of cyberware would have some number of "required stat points".  Most of them will require some Body.  So for example if your Character has 18 Body, the sum total of required Body points of all Cyberware implants cannot exceed 18.  If installing a piece of Cyberware would require more than 18 Body total for all Cyberware, then the Cyberware would not be possible to install.  This does not deduct the actual Body stat: the stat itself is just a limit on the amount of Cyberware you can install on your character.

Installing and removing cyberware would be done via a Cyberdoc Terminal like the Mecha Engineering Terminal, and bring up the Gear Editor as well. You can then install Cyberware into various parts of the body.  Caramel has a concept of "slot" for `sub_com` gears as well (`MULTIPLICITY_LIMITS`), which we can reuse: the Head of the character has a Eye, Ear, Skull, Brainstem slots, each of which can be fulfilled by some piece of Cyberware.  You can have different Cyberware in your arms --- perhaps your right arm has a Cyberware that improves Reflexes while your left arm has a Cybertware that improves Ranged Combat.

Trauma, like randomly destroying precious mecha equipment by installing them with low Mecha Engineering skill, would be left to previous GearHead games.  It's all about how much your Body can take.

Each piece of cyberware then indicates:

* Which body part (Head, Torso, Arm, Leg) it can be installed in.
* What slot on that body part it takes up.
* How many Body points it takes up.
* What bonuses it provides to stats.

Code-wise, our problem here is that the `Head`, `Torso`, etc. classes are currently coded strictly for mecha, and the `Character` reuses the mecha-centric object types.  This is fine for now because `Character`s cannot be edited in the gear editor currently, but it would be better to have separate `CharHead`, `CharTorso` etc. gear types that have `Character`-centric code; when loading a savegame we can look for `Character`s whose bodies are the mecha-centric gear types and have the `Character` rebuild the body with the proper types.  Alternately we can complicate `Head`, `Torso` etc. code with checks on the `self.scale` to separate mecha from characters, but that's fairly wonky IMO.

Lancemates might not be interested in letting you install or remove their cyberware, unless you have a very good relationship with them.

A piece of Cyberware would look like so:

```
Cyberware
    name = "Artificial Heart"
    part = CharTorso
    slot = "Heart"
    size = 6   % Body points taken up by this piece of Cyberware.
    statline = { Body = 2 }
```

Cyberware would of course be expensive.  And their infopanel should also display the above details.

Another possibility would be for Cyberware to also include a `programs` tuple so you can also get EW programs by installing them on your body, such as an AI Assistant or Target Analysis computer you can install in your Forebrain slot.

Head slots:

* Eye
* Ear
* Forebrain
* Brainstem

Torso slots:

* Heart
* Lung
* Spine

Arm slots:

* Muscle
* Bone

Leg slots:

* Muscle
* Bone

Cyberware would be made of Biotech, because srsly, and would require Biotechnology skill to repair, and would lose their provided benefit if destroyed.

Here's a bunch of cyberware ideas:

```
Cyberware
    name = "Microscope Eye"
    part = CharHead
    slot = "Eye"
    size = 9
    statline = { Science = 1 }
Cyberware
    name = "Targeter Eye"
    part = CharHead
    slot = "Eye"
    size = 9
    programs = (TargetAnalysis,)
Cyberware
    name = "Fashion Eyes"
    part = CharHead
    slot = "Eye"
    size = 6
    statline = { Charm = 2 }
Cyberware
    name = "In-head AI Assistant"
    part = CharHead
    slot = "Forebrain"
    size = 9
    programs = (AIAssistant,)
Cyberware
    name = "Data Link"
    part = CharHead
    slot = "Forebrain"
    size = 5
    statline = { Knowledge = 1 }
Cyberware
    name = "Predator Reflexes"
    part = CharHead
    slot = "Brainstem"
    size = 13
    statline = {Reflexes = 4, Perception = 3, Speed = 4, Ego = 1, Body = 1, Charm = -1, Craft = -1 }
Cyberware
    name = "Pilot Link"
    part = CharHead
    slot = "Brainstem"
    size = 10
    statline = { MechaPiloting = 2 }
```

Thoughts?  Feedback?  Is Cyberware fun?

**Comment by AmkG** [comment]
Hmm maybe we don't need to change the actual body parts.  Instead, we can hook into `is_legal_sub_com` and check the scale there.  We can make the cyberware location into different types and just add to the `MULTIPLICITY_LIMITS`, i.e.

```python
class MF_Torso( ModuleForm ):
    # blah
     MULTIPLICITY_LIMITS = {
          Engine: 1, Mount: 1, Cockpit: 1, Gyroscope: 1, Armor: 1
          EyeCyberware: 2, EarCyberware: 2, ForebrainCyberware: 1, BrainstemCyberware: 1
     }
```

Then we hook into `get_stat` of `Character`:

```python
class Character( Being ):
    # blah
    def get_stat(self, stat):
        value = super().get_stat(stat)
        for part in self.sub_com():
            for cyberware in part.sub_sub_com():
                if hasattr(cyberware, 'statline'):
                    value += cyberware.statline.get(stat, 0)
         return value
```

**Comment by AmkG** [comment]
Okay, so what I am working on now would look a little like this:

```
HeartCyberware
    name = "Cyber Heart"
    statline = { Body = 2 }
```

The amount of Body points needed is computed based on the `statline`.  The type of cyberware indicates where it can be installed.

**Comment by jwvhewitt** [comment]
Sounds good. One issue is that Cyberware should not be a no-brainer; in earlier GearHead games there were significant dangers of cybering up, but I don't really like constant maintenance and frequent trips to the hospital as a mechanism for GearHead Caramel*.

One option may be to reintroduce the Cyberware skill, with your skill value being how many points of cyberware you can install. The only ways to learn Cyberware would be to get it at character creation or to suffer a near-death event in-game.

*I'm trying to keep GHC easy to pick up and easy to put down again without the player doing a lot of maintenance or busywork. Hence the auto-repair and auto-pay at the end of missions, plus various other design decisions.This may be because I'm getting old and forgetful.

**Comment by AmkG** [comment]
> Cyberware should not be a no-brainer

My preferred design paradigm would be to give different interesting options and drawbacks.  Perhaps every piece of Cyberware should have at least -1 Body, for example, and Cyberware with +Body would have large reductions in other stats.  If all Cyberware are consistently reducing Body except a handful which have large reduction in stats (e.g. maybe Cyber Heart gives +1 Body but -2 Speed (because moving too fast may overstress the cyberheart, I dunno)) then your choice is either to make Body your dump stat in the first place and pump your other stats and not spend extra funds on Cyberware, OR pump Body and trade off points in Body and some funds for bonuses in stats.

In GearHead 1 the thing to do was to select Extropian and raise your Cybertech +4 and install Predator Reflexes, then pump your Cybertech and install various other Cyberware to taste.  We want to avoid that; Predator Reflexes in GHC would require preparing a character specifically that can take it.  Maybe we can give it -9999 Science -9999 Repair -9999 Medicine -9999 Biotechnology -9999 Computers -9999 Performance -9999 Negotiation, implying those skills will be lost if you have them, and everybody loves Epic Singing LISTEN TO MY SONG so nobody will use Predator Reflexes anymore.

> The only ways to learn Cyberware would be to get it at character creation or to suffer a near-death event in-game.

That's interesting as well.  So in case the player character dies from hitpoint loss zey could be revived, gain a few points of Cybertech, and lose some funds to pay for an emergency installation of some basic cyberware.

**Comment by AmkG** [comment]
In previous games once you install a piece of Cyberware it cannot be removed, only replaced.  My current thought is to allow Cyberware to be removed and installed arbitrarily, like mecha parts.  The Watsonian explanation is that technology has advanced enough that cyberdoc terminals can clone replacement organs if you eventually decide not to keep a piece of cyberware; the Doylist explanation is so that I don't have to design a new UI for Cyberware installation and just use the Gear Editor.

**Comment by AmkG** [comment]
"Isn't cyberware potentially dangerous?"

"You are thinking of the cyberware of yesteryear.  Recent regulations have tightened on cyberware manufacture and modern cyberware is free of side effects.  Of course, new cyberware still has drawbacks, such as modern Fashion Eyes reducing visual acuity, due to the necessary safety features now required to prevent cyberdisfunction.  Each piece of cyberware indicates the level of trauma they add, to a new standardized scale, and it is easy to assess if a piece of cyberware would put more load on your Body than it can handle."

"Okay, so how do I get cyberware installed?"

"The Cyberdoc Terminal removes human error during installation that was the root cause of many cyberdisfunctions.  Gone are the days of quack cyberdocs extracting massive payments for botched surgeries.  The new Cyberdoc Terminals even have a organ cloning capability, so you can remove cyberware entirely.  You can freely plan how much cyberware you will have installed or removed, then the Cyberdoc Terminal will perform the operation quickly and safely."

**Comment by AmkG** [comment]
I'm not quite convinced yet to implement a special Cybertech skill that gates access to cyberware, but I think there's a possibility of getting a near-death experience in the backstory (such as war or betrayal) that gives you a random installed cyberware.

"A stray blast hit a shelter you were hiding in, causing damage to your [location] and nearly killing you, but a [philantropic/heroic/atoning/guilty] cyberdoc performed emergency surgery to install [cyberware]."

"The traitor rigged a bomb in your cockpit that damaged your [location], and you were forced to undergo surgery to install [cyberware] to save your life."

**Comment by AmkG** [comment]
Okay, #81 got merged in.

Current rules:

* Max trauma is determined by your base unmodified Body stat.  You cannot install more cyberware than the max trauma your Body can take.
* Each point of trauma reduces stamina by one point.  Bonuses and maluses to Body, Ego, and Athletics change how much stamina a piece of cyberware will cost, but yeah, it's basically that.
* Anyone can get cyberware if they can buy and install it, no restrictions yet.
* Cyberware bonuses and maluses get summed up.
* Cyberware is purchased at the doctor in Wujung Hospital.
* Only the player character can install cyberware on themselves.
* Cyberware can be removed.

Stuff to debate about and maybe play with:

* Trauma (and the ability to gain cyberware) can be gated by a skill/badge you get by a near-death experience: if you do not have the skill/badge, your max trauma is 0, preventing cyberware usage.
  * Skill vs badge: is cyberware a continuous incremental improvement (Cybertech skill) or is it something which turns you into a fooming hard-takeoff singularity all by yourself (Extropian Badge)?
  * Neiither: whether to install cyberware or not is a choice that a person makes, nothing prevents anyone from e.g. using a computer or piloting a mecha, why should cyberware be so different?  Let the non-users of cyberware go on a shrine tour or whatever.
* Cyberware purchased directly at the cyberdoc terminal.
  * Mostly a matter of coding; it's more elegant that way.
* Only the largest bonus and largest malus is used
  * Or maybe only the largest bonus, maluses still sum up?  Otherwise you could select one stat you can afford a few points loss and install all the cyberware that give malus to them, with the maluses not stacking.
  * Not natural: GH1 and GH2 didn't work this way, many games with item-based bonuses will usually stack them instead of taking the max.
  * The elegant way to do this would be at the cyberdoc: when installing a new cyberware while you have an existing cyberware with the same bonus, before installing, warn the user and let them abort.
    * Best if purchasing cyberware is already done via the cyberdoc terminal.  That way you don't lose any funds if you only now found out that only the highest bonus (and highest malus?) is applied.
    * Can be annoying if you already know about it; suppress the warning if the character already have two or more cyberware that share bonuses to stats, the presumption is that the character already got warned before and already knows it.
* Lifepath that installs cyberware, so that lancemates can have cyberware and (if we gate cyberware with Cybertech skill or Extropian badge) possibly gain cyberware.
  * Generally, a "near-death experience" is what is planned: you got into a fatal accident, a passing heroic cyberdoc installs a cyberware for free and teaches you Cybertech / convinces you of the Extropian philosophy.
    * If cyberware can be removed, presumably that's because the cyberdoc can automatically regrow the organs that the cyberware replaced.  If you have the tech to regrow lost organs, then there is no real need to learn Cybertech / gain Extropian when a near-death experience causes fatal damage to your heart/spine/whatever: just use the same tech to rebuild the lost orgran without getting the cyberware.
    * Maybe some other lifepaths?
      * "I was an organ smuggler" for criminal life paths.
      * Cybertech could be a major course in University.
      * "Futurism" as another alternative to the Sciences or Arts for Autodidacts.

**Comment by AmkG** [comment]
Okay, looks like @jwvhewitt modified the trauma calculations, and the current rules are now:

* Max trauma is primarily dominated by Cybertech skill, but ordinary non-Cybertech characters can still get some cyberware (usually just one good one, or maybe two with a high Body and with the lower-trauma cyberware).

I'm actually quite fine with that, except for #88 which basically means it looks like some cyberware locations were not fully updated for the new trauma calcs.  I'm fine with non-Cybertech characters getting one cyberware, *maybe* two or three if they stretch.

**Comment by AmkG** [comment]
Okay, here are the rules as of the master branch of this writing:

* Max trauma is determined by your (unmodified) Body and Cybertech.  An average Body stat of 10 will give you a mere 3 points of trauma max (for reference, a Cyber Eye is 1 trauma, a Cyber Heart is 5 trauma).  One point of Cybertech is one extra point of trauma.
* Any character can have cyberware installed.  However the max trauma limit you get without Cybertech is very small; in practice you can install one good cyberware, or maybe two lesser ones with some effort/planning.
* You lose max stamina with cyberware installed.  1 trauma = 1 stamina lost.  This is in addition to any loss you get from losing Body, Ego, or Athletics.
* There are two lifepath options that give Cybertech: A black-market cyberdoc hard-knocks, or a course in university.
* Only the player character can enter the cyberdoc and have their cyberware modified.
* Can remove, replace, and install cyberware.
* You have two arms and two legs and can install separate cyberware in each (though there are no leg cyberware yet).
* Bonuses and maluses of multiple cyberware are summed up.
* Cyberware can be purchased (but not sold) directly at the cyberdoc.

I'm actually reasonably satisfied with the current ruleset.  So I'm closing this now.  I have a few more cyberware-related ideas but I'll just put them in separate new issues.

**Comment by jwvhewitt** [comment]
Yes, when I put in the Cyberware skill I threw in the trauma equation I think you suggested in the other thread. So now everyone can take a piece or two of cyberware, and specialists can take a lot more.

**Comment by AmkG** [comment]
Okay, so I just finished a playthrough with a Cybertech char.  The stamina reduction is indeed a good limit on the use of Cyberware; I actually had a mission where my mecha kept getting hit, necessitating multiple rounds of Repair, because I ran out of stamina (I had 11 points of trauma; there wasn't enough Reflexes bonus to go beyond that).  A few points of Athletics, though, was enough to fix that; I think we can bump up the stamina drain of cyberware a little more, maybe 1.1~1.2 stamina per trauma, rounded down (which amounts to one extra stamina point loss per 10~5 trauma points).  Or keep as is, maybe add more skills/attack attributes with stamina price (maybe Fast Attack can drain 1SP instead of 2MP? Stealth costs 2SP / 3SP instead of MP?  LISTEN TO MY SONG costs 4SP 4MP instead of 9MP (it's physically tiring to perform well, after all!)?)
